core/os/device/remotessh/commands.go

```diff
@@ -0,0 +1,246 @@
+		}
+	}
+
+	if err := session.Start(prefix + cmd.Name + " " + strings.Join(cmd.Args, " ")); err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (t sshShellTarget) String() string {
+	c := t.b.configuration
+	return c.User + "@" + c.Host + ": " + t.b.String()
+}
+
+// Shell implements the Device interface returning commands that will error if run.
+func (b binding) Shell(name string, args ...string) shell.Cmd {
+	return shell.Command(name, args...).On(sshShellTarget{&b})
+}
+
+func (b binding) destroyPosixDirectory(ctx context.Context, dir string) {
+	_, _ = b.Shell("rm", "-rf", dir).Call(ctx)
+}
+
+func (b binding) createPosixTempDirectory(ctx context.Context) (string, func(context.Context), error) {
+	dir, err := b.Shell("mktemp", "-d").Call(ctx)
+	if err != nil {
+		return "", nil, err
+	}
+	return dir, func(ctx context.Context) { b.destroyPosixDirectory(ctx, dir) }, nil
+}
+
+func (b binding) createWindowsTempDirectory(ctx context.Context) (string, func(ctx context.Context), error) {
+	return "", nil, fmt.Errorf("Not yet supported, windows remote targets")
+}
+
+func (b binding) MakeTempDir(ctx context.Context) (string, func(ctx context.Context), error) {
+	switch b.os {
+	case device.Linux:
+		fallthrough
+	case device.OSX:
+		return b.createPosixTempDirectory(ctx)
+	case device.Windows:
+		return b.createWindowsTempDirectory(ctx)
+	default:
+		panic("We should never end up here")
+	}
+}
+
+func (b binding) WriteFile(ctx context.Context, contents io.Reader, mode string, destPath string) error {
+	_, err := b.Shell("cat", ">", destPath, "; chmod ", mode, " ", destPath).Read(contents).Call(ctx)
+	return err
+}
+
+func (b binding) PushFile(ctx context.Context, source, dest string) error {
+	infile, err := os.Open(source)
+	if err != nil {
+		return err
+	}
+	permission, err := os.Stat(source)
+	if err != nil {
+		return err
+	}
+	perm := fmt.Sprintf("%4o", permission.Mode().Perm())
+
+	return b.WriteFile(ctx, infile, perm, dest)
+}
+
+func (b binding) doTunnel(ctx context.Context, local net.Conn, remotePort int) error {
+	remote, err := b.connection.Dial("tcp", fmt.Sprintf("localhost:%d", remotePort))
+	if err != nil {
+		local.Close()
+		return err
+	}
+
+	closer := make(chan bool)
+
+	copy := func(writer io.Writer, reader io.Reader) {
+		_, err := io.Copy(writer, reader)
+		if err != nil {
+			log.E(ctx, "Copy Error %s", err)
+		}
+		closer <- true
+	}
+
+	go copy(local, remote)
+	go copy(remote, local)
+	go func() {
+		defer local.Close()
+		defer remote.Close()
+		// When one direction of the communication has
+		// closed, close the entire connection
+		<-closer
+		<-closer
+	}()
+	return nil
+}
+
+func (b binding) ForwardPort(ctx context.Context, remotePort int) (int, error) {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return 0, err
+	}
+	go func() {
+		defer listener.Close()
+		for {
+			local, err := listener.Accept()
+			if err != nil {
+				return
+			}
+			go b.doTunnel(ctx, local, remotePort)
```

Comment:
I don't think this should be on a separate go-routine, as you're currently ignoring the error and nearly all the heavy work is offloaded to separate go-routines in `doTunnel` anyway.